pkg/k8s/service.go

```diff
@@ -91,12 +93,44 @@ func ParseService(svc *types.Service) (ServiceID, *Service) {
 	svcInfo.IncludeExternal = getAnnotationIncludeExternal(svc)
 	svcInfo.Shared = getAnnotationShared(svc)
 
-	// FIXME: Add support for
-	//  - NodePort
 	for _, port := range svc.Spec.Ports {
 		p := loadbalancer.NewFEPort(loadbalancer.L4Type(port.Protocol), uint16(port.Port))
-		if _, ok := svcInfo.Ports[loadbalancer.FEPortName(port.Name)]; !ok {
-			svcInfo.Ports[loadbalancer.FEPortName(port.Name)] = p
+		portName := loadbalancer.FEPortName(port.Name)
+		if _, ok := svcInfo.Ports[portName]; !ok {
+			svcInfo.Ports[portName] = p
+		}
+		// This is a hack;-( In the case of NodePort service, we need to create
+		// two surrogate frontends per IP protocol - one with a zero IP addr used
+		// by the host-lb, another with a public iface IP addr. For each frontend
+		// we will need to store a service ID used for a reverse NAT translation
+		// and for deleting a service.
+		// Unfortunately, doing this in daemon/{loadbalancer,k8s_watcher}.go
+		// would introduce more complexity in already too complex LB codebase,
+		// so for now (until we have refactored the LB code) keep NodePort
+		// frontends in Service.NodePorts.
+		if svc.Spec.Type == v1.ServiceTypeNodePort {
+			if option.Config.EnableNodePort {
+				if _, ok := svcInfo.NodePorts[portName]; !ok {
+					svcInfo.NodePorts[portName] =
+						make(map[string]*loadbalancer.L3n4AddrID)
+				}
+				proto := loadbalancer.L4Type(port.Protocol)
+				port := uint16(port.NodePort)
+				id := loadbalancer.ID(0) // will be allocated by k8s_watcher
+
```

Comment:
to avoid calculating the String() 4 times you can create a variable with the string already here `nodePortFEStr := nodePortFE.String()`

Reply:
As discussed offline, string is each time different (v4/v6 and wild-card/non-wild-card), so okay as is.